src/box/txn_limbo.c

```diff
@@ -773,38 +784,287 @@ txn_limbo_wait_empty(struct txn_limbo *limbo, double timeout)
+{
+	assert(latch_is_locked(&limbo->promote_latch));
+	assert(limbo->validate);
+
+	/*
+	 * Zero LSN is allowed for PROMOTE and DEMOTE requests only.
+	 */
+	if (req->lsn == 0) {
+		say_error("%s. Zero lsn detected", reject_str(req));
+		diag_set(ClientError, ER_UNSUPPORTED, "Replication",
+			 "zero LSN for CONFIRM/ROLLBACK");
+		return -1;
+	}
+
+	return filter_queue_boundaries(limbo, req);
+}
+
+/** A filter PROMOTE and DEMOTE packets. */
+static int
+filter_promote_demote(struct txn_limbo *limbo,
+		      const struct synchro_request *req)
+{
+	assert(latch_is_locked(&limbo->promote_latch));
+	assert(limbo->validate);
+
+	/*
+	 * PROMOTE and DEMOTE packets must not have zero
+	 * term supplied, otherwise it is a broken packet.
+	 */
+	if (req->term == 0) {
+		say_error("%s. Zero term detected", reject_str(req));
+		diag_set(ClientError, ER_UNSUPPORTED,
+			 "Replication", "PROMOTE/DEMOTE with a zero term");
+		return -1;
+	}
+
+	/*
+	 * If the term is already seen it means it comes
+	 * from a node which didn't notice new elections,
+	 * thus been living in subdomain and its data is
+	 * no longer consistent.
+	 */
+	if (limbo->promote_greatest_term >= req->term) {
+		say_error("%s. Max term seen is %llu", reject_str(req),
+			  (long long)limbo->promote_greatest_term);
+		diag_set(ClientError, ER_SPLIT_BRAIN,
+			 "got a PROMOTE/DEMOTE with an obsolete term");
+		return -1;
+	}
+
+	return filter_queue_boundaries(limbo, req);
+}
+
+/** A fine-grained filter checking specific request type constraints. */
+static int
+txn_limbo_filter_request(struct txn_limbo *limbo,
+			 const struct synchro_request *req)
+{
+	if (!limbo->validate)
+		return 0;
+	/*
+	 * Wait until all the entries receive an lsn. The lsn will be
+	 * used to determine whether filtered request is safe to apply.
+	 */
+	if (txn_limbo_wait_persisted(limbo) < 0)
+		return -1;
+	switch (req->type) {
+	case IPROTO_RAFT_CONFIRM:
+	case IPROTO_RAFT_ROLLBACK:
+		return filter_confirm_rollback(limbo, req);
+	case IPROTO_RAFT_PROMOTE:
+	case IPROTO_RAFT_DEMOTE:
+		return filter_promote_demote(limbo, req);
+	default:
+		unreachable();
+	}
+}
+
+int
 txn_limbo_req_prepare(struct txn_limbo *limbo,
 		      const struct synchro_request *req)
 {
 	assert(latch_is_locked(&limbo->promote_latch));
+
+	if (txn_limbo_filter_generic(limbo, req) < 0)
+		return -1;
+
+	/*
+	 * Guard against new transactions appearing during WAL write. It is
+	 * necessary because otherwise when PROMOTE/DEMOTE would be done and it
+	 * would see a txn without LSN in the limbo, it couldn't tell whether
+	 * the transaction should be confirmed or rolled back. It could be
+	 * delivered to the PROMOTE/DEMOTE initiator even before than to the
+	 * local TX thread, or could be not.
+	 *
+	 * CONFIRM and ROLLBACK need this guard only during  the filter stage.
+	 * Because the filter needs to see all the transactions LSNs to work
+	 * correctly.
+	 */
+	assert(!limbo->is_in_rollback);
+	limbo->is_in_rollback = true;
```

Comment:
Did I see correctly, that even if everything is fine and dandy, and we are writing a regular CONFIRM on the limbo owner itself, then new transactions will be rolled back on attempt to commit? Doesn't it look a bit hard to use if you can't safely write more than one synchro txn at a time or you will get rollbacks? Or did I misunderstand something?

Reply:
I re-checked, and it's fine.
Local CONFIRM and ROLLBACK bypass the filter completely. Just like they do with `txn_limbo_req_prepare`.

Reply:
Good, goooood.